V8intrinsic syntax plugin

JLHwung opened the pull request:
<!--
Before making a PR, please read our contributing guidelines
https://github.com/babel/babel/blob/master/CONTRIBUTING.md

For issue references: Add a comma-separated list of a [closing word](https://help.github.com/articles/closing-issues-via-commit-messages/) followed by the ticket number fixed by the PR. (it should be underlined in the preview if done correctly)

If you are making a change that should have a docs update: submit another PR to https://github.com/babel/website
-->

| Q                        | A <!--(Can use an emoji 👍) -->
| ------------------------ | ---
| Fixed Issues?            | Fixes #10104  <!-- remove the (`) quotes to link the issues -->
| Patch: Bug Fix?          |
| Major: Breaking Change?  |
| Minor: New Feature?      | ✔️
| Tests Added + Pass?      | ✔️
| Documentation PR Link    | <!-- If only readme change, add `[skip ci]` to your commits -->
| Any Dependency Changes?  | No
| License                  | MIT

<!-- Describe your changes below in as much detail as possible -->
Authored a new syntax plugin `v8intrinsic` to support V8 CallRuntime syntax
```
CallRuntime ::
  '%' Identifier Arguments
```

The original implementation in V8 parser `peek()` twice to check whether a `modulo` token is immediately followed by an identifier and a `parenL` token.

As in Babel `lookahead()` does not support multiple steps, here we implicitly track the execution branch inside `parseSubscript` to see if `parseCallExpressionArguments` is immediately called in the first round of `parseExprSubscripts`. If so we conclude that the modulo token is a valid V8Intrinsic marker.

It is also interesting to note that a V8Intrinsic call inside a _new_ expression does not raise a SyntaxError in V8. i.e.
```
node --allow-natives-syntax -e "new %DebugPrint(null)"
// throws TypeError: (intermediate value) is not a constructor

node -e "new %DebugPrint(null)"
// throws SyntaxError: Unexpected token %
```
@mathiasbynens I don't think there are any V8 intrinsics supposed to work with _new_ expression, aren't there?

If so, it should be reasonable to prevent v8intrisic from new expression, as is the behavior of this plugin. Otherwise we still need to take care on the _new_ expression.

### Action Items
- [x] Add `v8intrinsic` to babel-parser typings after #10259 gets merged.
- [x] Document updates: https://github.com/babel/website/pull/2074

JLHwung commented on packages/babel-parser/src/plugins/v8intrinsic.js:
```diff
@@ -0,0 +1,48 @@
+
+    createIdentifier(): N.Identifier {
+      const identifier = super.createIdentifier(...arguments);
+      if (this.state.V8IntrinsicStart !== undefined) {
```
Are there any way to document the type of `V8IntrinsicStart` be a `Number` or `Undefined` ?

nicolo-ribaudo replied:
You should specify it at https://github.com/babel/babel/blob/master/packages/babel-parser/src/tokenizer/state.js

Nit: I know that it's name is V8, but could you start this variable name with `v8`? :sweat_smile: We only use upper-first names for classes/interfaces.

JLHwung replied:
Good to know the naming convention here. Resolved as we have removed `V8IntrinsicStart` internal state.

nicolo-ribaudo commented on packages/babel-parser/test/fixtures/v8intrinsic/expression/await-expression/input.js:
```diff
@@ -0,0 +1 @@
+async () => { await %StringParseInt("42", 10) }
```
Can you also add a test for `yield`?

```js
function* fn() {
  yield %StringParseInt("42", 10);
}
```

We should ensure that it isn't parsed as
```js
function* fn() {
  (yield) % (StringParseInt("42", 10));
}
```

JLHwung replied:
Test case is added and passed.

JLHwung commented:
> I was thinking that maybe we can remove this.state.V8IntrinsicStart and put all the parsing logic in a single method

I agree, actually my first attack is on rewriting `parseIdentifier` but I find out it would mess up with the token position internal state. Apparently overriding `parseExprAtom` is way better than my previous approach. I have pushed the rewritten version.

> Can you also add a test for yield?

Great observation! The `parseYield` will check if a token can `startsExpr`, if not it ends up with an empty `yield` statement, which fails under this case where `%` can actually start an expression. I would try to figure out a way to do this in the plugin land.

mathiasbynens commented:
Thanks for your work on this!

> @mathiasbynens I don't think there are any V8 intrinsics supposed to work with new expression, aren't there?

That's correct. [Here's how V8 implements parsing support based on the flag.](https://cs.chromium.org/chromium/src/v8/src/parsing/parser-base.h?l=1785-1789&rcl=5e14bd7ec82b9021667402cc5c92310230bea48c)

mathiasbynens commented on packages/babel-parser/src/plugin-utils.js:
```diff
@@ -89,13 +96,15 @@ import flow from "./plugins/flow";
 import jsx from "./plugins/jsx";
 import typescript from "./plugins/typescript";
 import placeholders from "./plugins/placeholders";
+import v8instrinsic from "./plugins/v8intrinsic";
```
typo: `s/v8instrinsic/vv8intrinsic/`

JLHwung replied:
Done in a1a4558.

mathiasbynens commented on packages/babel-parser/test/fixtures/v8intrinsic/_errors/not-in-call-expression/input.js:
```diff
@@ -0,0 +1,2 @@
+const i = %DebugpPrint;
```
typo: `%DebugPrint`

(not that it really matters for the purposes of the test case)

JLHwung replied:
Ahh, good catch. Fixed and pushed.

mathiasbynens commented on packages/babel-parser/test/fixtures/v8intrinsic/options.json:
```diff
@@ -0,0 +1,3 @@
+{
+  "plugins": ["v8instrinsic"]
```
typo: `v8intrinsic`

JLHwung replied:
Done in a1a4558.

JLHwung commented:
> That's correct. Here's how V8 implements parsing support based on the flag.

The implementation of V8 parsing `V8Intrinsic` will lead to the fact that `new %StringParseInt("42", 10)` will _not_ throw a syntax error, instead it throws a runtime type error. 

However as I have rewritten the plugin and now it closely follow the way how V8 parse. I have added this `newExpression` test case to the success test suites.

JLHwung commented:
I've talked to Nicolo offline a bit. Here we add `{ startsExpr: true }` configuration to `parser/types` since

- It is only used in `parseYield`. The impact is under control.
- There have been some plugin-defined tokens inside this file, i.e. `braceBarL`, `braceBarR`. It is acceptable to place plugin-defined token configurations here.

jridgewell commented:
@nicolo-ribaudo: Is this waiting for a release, or can it be merged?

nicolo-ribaudo commented:
We just released v7.5 so this is waiting for the next minor, but if it is urgent we could do it earlier than the usual time between two minor releases.

nicolo-ribaudo commented on packages/babel-types/src/definitions/misc.js:
```diff
@@ -19,3 +23,12 @@ defineType("Placeholder", {
   },
 });
+
+defineType("V8IntrinsicIdentifier", {
```
I just realized that we should add this to the allowed `callee`s of `CallExpression`
